packages/utils/src/libp2p/dialHelper.ts

```diff
@@ -49,89 +50,217 @@ export type DialResponse =
+      break
+    }
+  }
+
+  if (deadConnections.length > 0) {
+    log(
+      `dead connection${deadConnections.length == 1 ? '' : 's'} to ${destination.toB58String()}`,
+      deadConnections.map((conn: Connection) => conn.id)
+    )
+  }
+
+  for (const deadConnection of deadConnections) {
+    libp2p.connectionManager.onDisconnect(deadConnection)
+  }
+
+  if (stream != undefined && conn != undefined) {
+    return { conn, stream, protocol }
+  }
+}
+
 /**
  * Performs a dial attempt and handles possible errors.
  * @param libp2p Libp2p instance
  * @param destination which peer to dial
  * @param protocol which protocol to use
  * @param opts timeout options
  */
-async function attemptDial(
-  libp2p: Pick<LibP2P, 'dialProtocol'>,
+export async function establishNewConnection(
+  libp2p: Pick<ReducedLibp2p, 'peerStore' | 'dial'>,
   destination: PeerId,
   protocol: string,
   opts: Required<TimeoutOpts>
-): Promise<
-  | DialResponse
-  | {
-      status: InternalDialStatus.CONTINUE
-    }
-> {
+): Promise<{
+  stream: MuxedStream
+  conn: Connection
+  protocol: string
+} | void> {
+  const knownAddresses = await libp2p.peerStore.addressBook.get(destination)
+
+  if (knownAddresses.length == 0) {
+    return
+  }
+
   const start = Date.now()
-  let struct: Awaited<ReturnType<LibP2P['dialProtocol']>> | null
 
   let aborted = false
 
   const onAbort = () => {
     aborted = true
   }
+
   opts.signal.addEventListener('abort', onAbort)
 
+  let conn: Connection
   try {
-    struct = await libp2p.dialProtocol(destination, protocol, { signal: opts.signal })
+    conn = await libp2p.dial(destination, { signal: opts.signal })
   } catch (err) {
     logError(`Error while dialing ${destination.toB58String()} directly.`)
     if (err?.message) {
       logError(`Dial error:`, err)
     }
   }
 
+  if (!conn) {
+    return
+  }
+
+  const stream = (await timeout(1000, () => conn.newStream(protocol)))?.stream
+
   opts.signal.removeEventListener('abort', onAbort)
 
   // Libp2p's return types tend to change every now and then
-  if (struct != null) {
-    if (aborted) {
-      console.log(`ending obsolete write stream after ${Date.now() - start} ms`)
-      try {
-        struct.stream
-          .sink((async function* () {})())
-          .catch((err: any) => logError(`Error while ending obsolete write stream`, err))
-      } catch (err) {
-        logError(`Error while ending obsolete write stream`, err)
-      }
+  if (stream != null && aborted) {
+    log(`ending obsolete write stream after ${Date.now() - start} ms`)
+    try {
+      stream
+        .sink((async function* () {})())
+        .catch((err: any) => logError(`Error while ending obsolete write stream`, err))
+    } catch (err) {
+      logError(`Error while ending obsolete write stream`, err)
     }
-    return { status: DialStatus.SUCCESS, resp: struct }
+    return
+  }
+
+  if (!stream) {
+    return
+  }
+
+  return {
+    conn,
+    stream,
+    protocol
+  }
```

Comment:
Consider making this a function, since the exact same code fragment is also at lines 234-263.

Reply:
changed